crypto/dllmain.c

```diff
@@ -25,34 +25,55 @@
  * detaches
  */
 
+static CRITICAL_SECTION csThreadStop;
+static DWORD stop_key;
+static void (*stop_fun)(void *);
+
+int CRYPTO_THREAD_register_stop_function(DWORD key, void (*cleanup)(void *))
+{
+    if (cleanup == NULL)
+        return 1;
+
+    EnterCriticalSection(&csThreadStop);
+    if (stop_fun != NULL) {
+        /* We handle only one stop function. */
+        LeaveCriticalSection(&csThreadStop);
+        return 0;
+    }
+
+    stop_key = key;
+    stop_fun = cleanup;
+    LeaveCriticalSection(&csThreadStop);
+    return 1;
+}
+
+void CRYPTO_THREAD_unregister_stop_function(DWORD key)
+{
+    EnterCriticalSection(&csThreadStop);
+    if (key == stop_key) {
+        stop_key = 0;
+        stop_fun = NULL;
+    }
+    LeaveCriticalSection(&csThreadStop);
+}
+
 BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved);
 BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
 {
     switch (fdwReason) {
     case DLL_PROCESS_ATTACH:
-        OPENSSL_cpuid_setup();
-# if defined(_WIN32_WINNT)
-        {
-            IMAGE_DOS_HEADER *dos_header = (IMAGE_DOS_HEADER *) hinstDLL;
-            IMAGE_NT_HEADERS *nt_headers;
-
-            if (dos_header->e_magic == IMAGE_DOS_SIGNATURE) {
-                nt_headers = (IMAGE_NT_HEADERS *) ((char *)dos_header
-                                                   + dos_header->e_lfanew);
-                if (nt_headers->Signature == IMAGE_NT_SIGNATURE &&
-                    hinstDLL !=
-                    (HINSTANCE) (nt_headers->OptionalHeader.ImageBase))
-                    OPENSSL_NONPIC_relocated = 1;
-            }
-        }
-# endif
+        InitializeCriticalSection(&csThreadStop);
         break;
     case DLL_THREAD_ATTACH:
         break;
     case DLL_THREAD_DETACH:
-        OPENSSL_thread_stop();
+        EnterCriticalSection(&csThreadStop);
+        if (stop_fun != NULL)
+            (*stop_fun)(TlsGetValue(stop_key));
+        LeaveCriticalSection(&csThreadStop);
```

Comment:
I don't quite get what is achieved. What apparently happens is that you record *first* destructor passed to CRYPTO_THREAD_init_local and call it upon detach. This calls for several questions. Why does first one gets so special treatment? Is there guarantee that first one is the "right" one? What about others? If all this relies on something utterly specific, then it would be appropriate to add comment.

Reply:
Well, the first  CRYPTO_THREAD_init_local has to be the one in ossl_init_base
that is why I had to invent the OPENSSL_INIT_BASE_ONLY. As you can see
it is the only one with a callback function, and all others don't have a callback
function.
Trouble is that it's infeasible to write all the background, and there always be
something that is implied to be known.

Reply:
> :-) But on serious note, I didn't actually deny the idea of a minimalistic initial pointer.

Yes, and that is why I wrote the comment `/* We handle only one stop function. */`
in the CRYPTO_THREAD_register_stop_function.

> What I'm saying is that it might be more appropriate to harmonize behaviour across platforms...

Yes, however this will probably only ever cover the case where OpenSSL is built as a DLL.
On the other side, on posix the callback will be called, but _after_ OpenSSL thread exit, and
using any OpenSSL functions in the callback can easily create a memory leak.

So using this feature is just a bug waiting to happen.

A quick check confirms, that I am right:
```
$ git diff .
diff --git a/test/threadstest.c b/test/threadstest.c
index ee09f86..f356e51 100644
--- a/test/threadstest.c
+++ b/test/threadstest.c
@@ -12,6 +12,7 @@
 #endif
 
 #include <openssl/crypto.h>
+#include <openssl/rand.h>
 #include "testutil.h"
 
 #if !defined(OPENSSL_THREADS) || defined(CRYPTO_TDEBUG)
@@ -132,6 +133,8 @@ static void thread_local_destructor(void *arg)
     count = arg;
 
     (*count)++;
+    int test, x=RAND_bytes((unsigned char*)&test, sizeof(test));
+    printf("x=%d test=%x\n", x, test);
 }
 
 static void thread_local_thread_cb(void)
$ LD_LIBRARY_PATH=..:$LD_LIBRARY_PATH ./threadstest
1..3
ok 1 - test_lock
ok 2 - test_once
x=1 test=5cdab57a
ok 3 - test_thread_local

=================================================================
==11477==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 352 byte(s) in 1 object(s) allocated from:
    #0 0x7fb4d43da800 in __interceptor_malloc ../../../../gcc-trunk/libsanitizer/asan/asan_malloc_linux.cc:86
    #1 0x7fb4d3df2130 in CRYPTO_secure_zalloc crypto/mem_sec.c:140
    #2 0x7fb4d3e5a18f in rand_drbg_new crypto/rand/drbg_lib.c:175
    #3 0x7fb4d3e5a18f in RAND_DRBG_secure_new crypto/rand/drbg_lib.c:243
    #4 0x7fb4d3e5b060 in drbg_setup crypto/rand/drbg_lib.c:853
    #5 0x7fb4d3e5b060 in RAND_DRBG_get0_public crypto/rand/drbg_lib.c:1031
    #6 0x7fb4d3e5b0f0 in drbg_bytes crypto/rand/drbg_lib.c:941
    #7 0x40323b in thread_local_destructor test/threadstest.c:136
    #8 0x7fb4d367ff81 in __nptl_deallocate_tsd (/lib/x86_64-linux-gnu/libpthread.so.0+0x7f81)

Indirect leak of 264 byte(s) in 1 object(s) allocated from:
    #0 0x7fb4d43da800 in __interceptor_malloc ../../../../gcc-trunk/libsanitizer/asan/asan_malloc_linux.cc:86
    #1 0x7fb4d3df0630 in CRYPTO_zalloc crypto/mem.c:230
    #2 0x7fb4d3dc9f50 in EVP_CipherInit_ex crypto/evp/evp_enc.c:128
    #3 0x7fb4d3e57ae9 in drbg_ctr_init crypto/rand/drbg_ctr.c:416
    #4 0x7fb4d3e57ec0 in RAND_DRBG_set crypto/rand/drbg_lib.c:122
    #5 0x7fb4d3e5a2f0 in rand_drbg_new crypto/rand/drbg_lib.c:208
    #6 0x7fb4d3e5a2f0 in RAND_DRBG_secure_new crypto/rand/drbg_lib.c:243
    #7 0x7fb4d3e5b060 in drbg_setup crypto/rand/drbg_lib.c:853
    #8 0x7fb4d3e5b060 in RAND_DRBG_get0_public crypto/rand/drbg_lib.c:1031
    #9 0x7fb4d3e5b0f0 in drbg_bytes crypto/rand/drbg_lib.c:941
    #10 0x40323b in thread_local_destructor test/threadstest.c:136
    #11 0x7fb4d367ff81 in __nptl_deallocate_tsd (/lib/x86_64-linux-gnu/libpthread.so.0+0x7f81)

Indirect leak of 264 byte(s) in 1 object(s) allocated from:
    #0 0x7fb4d43da800 in __interceptor_malloc ../../../../gcc-trunk/libsanitizer/asan/asan_malloc_linux.cc:86
    #1 0x7fb4d3df0630 in CRYPTO_zalloc crypto/mem.c:230
    #2 0x7fb4d3dc9f50 in EVP_CipherInit_ex crypto/evp/evp_enc.c:128
    #3 0x7fb4d3e56626 in ctr_update crypto/rand/drbg_ctr.c:274
    #4 0x7fb4d3e575eb in drbg_ctr_generate crypto/rand/drbg_ctr.c:349
    #5 0x7fb4d3e598ca in RAND_DRBG_generate crypto/rand/drbg_lib.c:626
    #6 0x7fb4d3e59e02 in RAND_DRBG_bytes crypto/rand/drbg_lib.c:658
    #7 0x40323b in thread_local_destructor test/threadstest.c:136
    #8 0x7fb4d367ff81 in __nptl_deallocate_tsd (/lib/x86_64-linux-gnu/libpthread.so.0+0x7f81)

Indirect leak of 168 byte(s) in 1 object(s) allocated from:
    #0 0x7fb4d43da800 in __interceptor_malloc ../../../../gcc-trunk/libsanitizer/asan/asan_malloc_linux.cc:86
    #1 0x7fb4d3df0630 in CRYPTO_zalloc crypto/mem.c:230
    #2 0x7fb4d3e57d44 in drbg_ctr_init crypto/rand/drbg_ctr.c:412
    #3 0x7fb4d3e57ec0 in RAND_DRBG_set crypto/rand/drbg_lib.c:122
    #4 0x7fb4d3e5a2f0 in rand_drbg_new crypto/rand/drbg_lib.c:208
    #5 0x7fb4d3e5a2f0 in RAND_DRBG_secure_new crypto/rand/drbg_lib.c:243
    #6 0x7fb4d3e5b060 in drbg_setup crypto/rand/drbg_lib.c:853
    #7 0x7fb4d3e5b060 in RAND_DRBG_get0_public crypto/rand/drbg_lib.c:1031
    #8 0x7fb4d3e5b0f0 in drbg_bytes crypto/rand/drbg_lib.c:941
    #9 0x40323b in thread_local_destructor test/threadstest.c:136
    #10 0x7fb4d367ff81 in __nptl_deallocate_tsd (/lib/x86_64-linux-gnu/libpthread.so.0+0x7f81)

Indirect leak of 168 byte(s) in 1 object(s) allocated from:
    #0 0x7fb4d43da800 in __interceptor_malloc ../../../../gcc-trunk/libsanitizer/asan/asan_malloc_linux.cc:86
    #1 0x7fb4d3df0630 in CRYPTO_zalloc crypto/mem.c:230
    #2 0x7fb4d3e57bdb in drbg_ctr_init crypto/rand/drbg_ctr.c:396
    #3 0x7fb4d3e57ec0 in RAND_DRBG_set crypto/rand/drbg_lib.c:122
    #4 0x7fb4d3e5a2f0 in rand_drbg_new crypto/rand/drbg_lib.c:208
    #5 0x7fb4d3e5a2f0 in RAND_DRBG_secure_new crypto/rand/drbg_lib.c:243
    #6 0x7fb4d3e5b060 in drbg_setup crypto/rand/drbg_lib.c:853
    #7 0x7fb4d3e5b060 in RAND_DRBG_get0_public crypto/rand/drbg_lib.c:1031
    #8 0x7fb4d3e5b0f0 in drbg_bytes crypto/rand/drbg_lib.c:941
    #9 0x40323b in thread_local_destructor test/threadstest.c:136
    #10 0x7fb4d367ff81 in __nptl_deallocate_tsd (/lib/x86_64-linux-gnu/libpthread.so.0+0x7f81)

SUMMARY: AddressSanitizer: 1216 byte(s) leaked in 5 allocation(s).
```